File: rofunc/planning/src/robotics-codes-from-scratch-master/matlab/LQT_CP_DMP.m
%%    Linear quadratic tracking (LQT) with control primitives applied to a trajectory 
%%    tracking task, with a formulation similar to dynamical movement primitives (DMP),
%%    by using the least squares formulation of recursive LQR on an augmented state space 
%%
%%    Copyright (c) 2021 Idiap Research Institute, http://www.idiap.ch/
%%    Written by Sylvain Calinon <https://calinon.ch>
%%
%%    This file is part of RCFS.
%%
%%    RCFS is free software: you can redistribute it and/or modify
%%    it under the terms of the GNU General Public License version 3 as
%%    published by the Free Software Foundation.
%%
%%    RCFS is distributed in the hope that it will be useful,
%%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
%%    GNU General Public License for more details.
%%
%%    You should have received a copy of the GNU General Public License
%%    along with RCFS. If not, see <http://www.gnu.org/licenses/>.

function LQT_CP_DMP

%% Parameters
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
param.nbData = 100; %Number of datapoints
param.nbSamples = 10; %Number of generated trajectory samples
param.nbVarU = 2; %Dimension of position data (here: x1,x2)
param.nbDeriv = 3; %Number of static and dynamic features (nbDeriv=2 for [x,dx] and u=ddx)
param.nbVar = param.nbVarU * param.nbDeriv; %Dimension of state vector
param.nbVarX = param.nbVar + 1; %Augmented state space
param.dt = 1E-2; %Time step duration
param.nbFct = 12; %Number of basis functions
param.basisName = 'RBF'; %PIECEWISE, RBF, BERNSTEIN, FOURIER
param.r = 1E-9; %Control cost in LQR 

%Dynamical System settings (for augmented state space)
A1d = zeros(param.nbDeriv);
for i=0:param.nbDeriv-1
	A1d = A1d + diag(ones(param.nbDeriv-i,1),i) * param.dt^i / factorial(i); %Discrete 1D
end
B1d = zeros(param.nbDeriv,1); 
for i=1:param.nbDeriv
	B1d(param.nbDeriv-i+1) = param.dt^i / factorial(i); %Discrete 1D
end
A0 = kron(A1d, eye(param.nbVarU)); %Discrete nD
B0 = kron(B1d, eye(param.nbVarU)); %Discrete nD
A = [A0, zeros(param.nbVar,1); zeros(1,param.nbVar), 1]; %Augmented A (homogeneous)
B = [B0; zeros(1,param.nbVarU)]; %Augmented B (homogeneous)

%Build Sx and Su transfer matrices (for augmented state space)
Sx = kron(ones(param.nbData,1), speye(param.nbVarX));
Su = sparse(param.nbVarX * param.nbData, param.nbVarU * (param.nbData-1));
M = B;
for t=2:param.nbData
	id1 = (t-1)*param.nbVarX+1:param.nbData*param.nbVarX;
	Sx(id1,:) = Sx(id1,:) * A;
	id1 = (t-1)*param.nbVarX+1:t*param.nbVarX; 
	id2 = 1:(t-1)*param.nbVarU;
	Su(id1,id2) = M;
	M = [A*M(:,1:param.nbVarU), M]; 
end

%Build basis functions
if isequal(param.basisName,'PIECEWISE')
	phi = buildPhiPiecewise(param.nbData-1, param.nbFct);
elseif isequal(param.basisName,'RBF')
	phi = buildPhiRBF(param.nbData-1, param.nbFct);
elseif isequal(param.basisName,'BERNSTEIN')
	phi = buildPhiBernstein(param.nbData-1, param.nbFct);
elseif isequal(param.basisName,'FOURIER')
	phi = buildPhiFourier(param.nbData-1, param.nbFct);
end

%Application of basis functions to multidimensional control commands
Psi = kron(phi, eye(param.nbVarU)); 


%% Task description
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
demos=[];
load('../data/2Dletters/S.mat'); %Planar trajectories for writing alphabet letters
MuPos = spline(1:size(demos{1}.pos,2), demos{1}.pos, linspace(1,size(demos{1}.pos,2),param.nbData)); %Position
MuVel = gradient(MuPos) / param.dt;
MuAcc = gradient(MuVel) / param.dt;
Mu = [MuPos; MuVel; MuAcc; zeros(param.nbVar-3*param.nbVarU, param.nbData)]; %Position, velocity and acceleration profiles as references

%Task setting (tracking of acceleration profile and reaching of an end-point)
Q = kron(speye(param.nbData), ...
    diag([zeros(1,param.nbVarU*2), ones(1,param.nbVarU) * 1E-6, zeros(1,param.nbVar-3*param.nbVarU)])); %Precision matrix (for acceleration only)
Q(end-param.nbVar+1:end-param.nbVar+2*param.nbVarU, end-param.nbVar+1:end-param.nbVar+2*param.nbVarU) = eye(2*param.nbVarU) * 1E0; %Add last point as target (position and velocity)

%Weighting matrices in augmented state form
Qm = zeros(param.nbVarX * param.nbData);
for t=1:param.nbData
	id0 = [1:param.nbVar] + (t-1) * param.nbVar;
	id = [1:param.nbVarX] + (t-1) * param.nbVarX;
	Qm(id,id) = [eye(param.nbVar), zeros(param.nbVar,1); -Mu(:,t)', 1] * blkdiag(Q(id0,id0), 1) * ...
	            [eye(param.nbVar), -Mu(:,t); zeros(1,param.nbVar), 1];
end
Rm = speye((param.nbData-1)*param.nbVarU) * param.r; %Standard control weight matrix (at trajectory level)


%% Least squares formulation of recursive LQR with an augmented state space and and control primitives
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%xn = [randn(param.nbVarU, 1) * 5E0; randn(param.nbVar-param.nbVarU, 1) * 0; 0]; %Simulated noise on state
xn = [3; -.5; zeros(param.nbVarX-param.nbVarU, 1)]; %Simulated noise on state

%F = (Su' * Qm * Su + Rm) \ Su' * Qm * Sx; %Standard F
W = (Psi' * Su' * Qm * Su * Psi + Psi' * Rm * Psi) \ Psi' * Su' * Qm * Sx;
F = Psi * W; %F with control primitives

%Reproduction with feedback controller on augmented state space (with CP)
Ka(:,:,1) = F(1:param.nbVarU,:);
P = eye(param.nbVarX);
for t=2:param.nbData-1
	id = (t-1)*param.nbVarU + [1:param.nbVarU];
	P = P / (A - B * Ka(:,:,t-1));
	Ka(:,:,t) = F(id,:) * P;
end
for n=1:2
	x = [Mu(:,1) + [2; 1; zeros(param.nbVar-2,1)]; 1]; %Augmented state space initialization (by simulating position offset)
	for t=1:param.nbData-1
		u = -Ka(:,:,t) * x; %Feedback control on augmented state (resulting in feedback and feedforward terms on state)
%		K = Ka(:,1:param.nbVar,t); %Feedback gain
%		uff = -Ka(:,end,t) - K * Mu(:,t); %Feedforward control command
%		u = K * (Mu(:,t) - x(1:end-1)) + uff; %Acceleration command with feedback and feedforward terms computed explicitly from Ka

		x = A * x + B * u; %Update of state vector	
		if t==25 && n==2
			x = x + xn; %Simulated noise on the state
		end
		r(n).x(:,t) = x; %State
	end
end


%% Plot 2D
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
h(1) = figure('position',[10 10 800 800]); hold on; axis off;
plot(Mu(1,:), Mu(2,:), 'b-','linewidth',2);
plot(Mu(1,end), Mu(2,end), 'r.','markersize',40);
plot(r(1).x(1,1), r(1).x(2,1), 'k.','markersize',30);
plot(r(1).x(1,:), r(1).x(2,:), 'k:','linewidth',2);
plot(r(2).x(1,:), r(2).x(2,:), 'k-','linewidth',2);
plot(r(2).x(1,24:25), r(2).x(2,24:25), 'g.','markersize',20);
axis equal; 

waitfor(h);
end


%Building piecewise constant basis functions
function phi = buildPhiPiecewise(nbData, nbFct) 
	phi = kron(eye(nbFct), ones(ceil(nbData/nbFct),1));
	phi = phi(1:nbData,:);
end

%Building radial basis functions (RBFs)
function phi = buildPhiRBF(nbData, nbFct) 
	t = linspace(0, 1, nbData);
	tMu = linspace(t(1)-1/(nbFct-3), t(end)+1/(nbFct-3), nbFct); %Repartition of centers to limit border effects
	sigma = 1 / (nbFct-2); %Standard deviation
	phi = exp(-(t' - tMu).^2 / sigma^2);
	
	%Optional rescaling
	%phi = phi ./ repmat(sum(phi,2), 1, nbFct); 
end

%Building Bernstein basis functions
function phi = buildPhiBernstein(nbData, nbFct)
	t = linspace(0, 1, nbData);
	phi = zeros(nbData, nbFct);
	for i=1:nbFct
		phi(:,i) = factorial(nbFct-1) ./ (factorial(i-1) .* factorial(nbFct-i)) .* (1-t).^(nbFct-i) .* t.^(i-1);
	end
end

%Building Fourier basis functions
function phi = buildPhiFourier(nbData, nbFct)
	t = linspace(0, 1, nbData);
	
	%Computation for general signals (incl. complex numbers)
	d = ceil((nbFct-1)/2);
	k = -d:d;
	phi = exp(t' * k * 2 * pi * 1i); 
	%phi = cos(t' * k * 2 * pi); %Alternative computation for real signal
		
%	%Alternative computation for real and even signal
%	k = 0:nbFct-1;
%	phi = cos(t' * k * 2 * pi);
%	%phi(:,2:end) = phi(:,2:end) * 2;
%	%invPhi = cos(k' * t * 2 * pi) / nbData;
%	%invPsi = kron(invPhi, eye(param.nbVar));
end

File: rofunc/planning/src/robotics-codes-from-scratch-master/python/LQT_CP_DMP.py
'''
    Linear quadratic tracking (LQT) with control primitives applied to a trajectory
    tracking task, with a formulation similar to dynamical movement primitives (DMP),
    by using the least squares formulation of recursive LQR on an augmented state space

    Copyright (c) 2021 Idiap Research Institute, http://www.idiap.ch/
    Written by Sylvain Calinon <https://calinon.ch>
    Written by Boyang Ti <https://www.idiap.ch/~bti/>

    This file is part of RCFS.

    RCFS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 3 as
    published by the Free Software Foundation.

    RCFS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with RCFS. If not, see <http://www.gnu.org/licenses/>.
'''
import os 
from pathlib import Path

import numpy as np
import matplotlib.pyplot as plt
from scipy import special
from math import factorial

import scipy.io as scio
from scipy.interpolate import interp1d
from scipy.linalg import block_diag

# Building piecewise constant basis functions
def build_phi_piecewise(nb_data, nb_fct):
    phi = np.kron( np.identity(nb_fct) , np.ones((int(np.ceil(nb_data/nb_fct)),1)) )
    return phi[:nb_data]

# Building radial basis functions (RBFs)
def build_phi_rbf(nb_data, nb_fct):
    t = np.linspace(0,1,nb_data).reshape((-1,1))
    tMu = np.linspace( t[0] , t[-1] , nb_fct )
    phi = np.exp( -1e2 * (t.T - tMu)**2 )
    return phi.T

# Building Bernstein basis functions
def build_phi_bernstein(nb_data, nb_fct):
    t = np.linspace(0,1,nb_data)
    phi = np.zeros((nb_data,nb_fct))
    for i in range(nb_fct):
        phi[:,i] = factorial(nb_fct-1) / (factorial(i) * factorial(nb_fct-1-i)) * (1-t)**(nb_fct-1-i) * t**i
    return phi

# Building Fourier basis functions
def build_phi_fourier(nb_data, nb_fct):

    t = np.linspace(0,1,nb_data).reshape((-1,1))

    # Alternative computation for real and even signal
    k = np.arange(0,nb_fct).reshape((-1,1))
    phi = np.cos( t.T * k * 2 * np.pi )
    return phi.T

# General param parameters
# ===============================
# Parameters
class Param:
    def __init__(self):
        self.dt = 1e-2  # Time step length
        self.nbData = 100  # Number of datapoints
        self.nbSamples = 10  # Number of generated trajectory samples
        self.nbVarU = 2  # Control space dimension (dx1,dx2,dx3)
        self.nbFct = 9  # Number of basis function
        self.nbDeriv = 3  # Number of static and dynamic features (nbDeriv=2 for [x,dx] and u=ddx)
        self.r = 1e-9  # Control weight term
        self.basisName = "RBF"  # can be PIECEWEISE, RBF, BERNSTEIN, FOURIER


# Main program
# ===============================
param = Param()


nb_var = param.nbVarU * param.nbDeriv  # Dimension of state vector
param.nbVar = nb_var  # Dimension of state vector
nb_varX = nb_var + 1
param.nbVarX = nb_varX  # Augmented state space


# Dynamical System settings (for augmented state space)
# =====================================
A1d = np.zeros(param.nbDeriv)
for i in range(param.nbDeriv):
    A1d = A1d + np.diag(np.ones((1, param.nbDeriv - i)).flatten(), i) * param.dt ** i * 1 / factorial(i)  # Discrete 1D

B1d = np.zeros((param.nbDeriv, 1))
for i in range(param.nbDeriv):
    B1d[param.nbDeriv - 1 - i] = param.dt ** (i + 1) * 1 / factorial(i + 1)  # Discrete 1D

A0 = np.kron(A1d, np.eye(param.nbVarU))  # Discrete nD
B0 = np.kron(B1d, np.eye(param.nbVarU))  # Discrete nD

A = np.vstack((np.hstack((A0, np.zeros((param.nbVar, 1)))), np.hstack((np.zeros((param.nbVar)), 1)).reshape(1, -1)))  # Augmented A (homogeneous)
B = np.vstack((B0, np.zeros((1, param.nbVarU))))  # Augmented B (homogeneous)

# Build Sx and Su transfer matrices (for augmented state space)
Sx = np.kron(np.ones((param.nbData, 1)), np.eye(param.nbVarX, param.nbVarX))
Su = np.zeros((param.nbVarX*param.nbData, param.nbVarU * (param.nbData-1)))  # It's maybe n-1 not sure
M = B
for i in range(1, param.nbData):
    Sx[i*param.nbVarX:param.nbData*param.nbVarX, :] = np.dot(Sx[i*param.nbVarX:param.nbData*param.nbVarX, :], A)
    Su[param.nbVarX*i:param.nbVarX*i+M.shape[0], 0:M.shape[1]] = M
    M = np.hstack((np.dot(A, M), B))  # [0,nb_state_var-1]


# Building basis funcitons
# =========================
functions = {
    "PIECEWEISE": build_phi_piecewise,
    "RBF": build_phi_rbf,
    "BERNSTEIN": build_phi_bernstein,
    "FOURIER": build_phi_fourier
}
phi = functions[param.basisName](param.nbData-1, param.nbFct)

# Application of basis functions to multidimensional control commands
PSI = np.kron(phi, np.identity(param.nbVarU))

# Task description
# =====================================

FILE_PATH = os.path.dirname(os.path.realpath(__file__))
DATASET_LETTER_FILE = "../data/2Dletters/S.npy"
x = np.load(str(Path(FILE_PATH, DATASET_LETTER_FILE)))[0,:,:2].T

f_pos = interp1d(np.linspace(0, np.size(x, 1)-1, np.size(x, 1), dtype=int), x, kind='cubic')
MuPos = f_pos(np.linspace(0, np.size(x, 1)-1, param.nbData))  # Position
MuVel = np.gradient(MuPos)[1] / param.dt
MuAcc = np.gradient(MuVel)[1] / param.dt

# Position, velocity and acceleration profiles as references
Mu = np.vstack((MuPos, MuVel, MuAcc, np.zeros((param.nbVar-3*param.nbVarU, param.nbData))))

# Task setting (tracking of acceleration profile and reaching of an end-point)
Q = np.kron(np.identity(param.nbData), np.diag(np.concatenate((np.zeros((param.nbVarU*2)), np.ones(param.nbVarU)*1e-6))))

Q[-1-param.nbVar+1:-1-param.nbVar+2*param.nbVarU+1, -1-param.nbVar+1:-1-param.nbVar+2*param.nbVarU+1] = np.identity(2*param.nbVarU) * 1e0

# Weighting matrices in augmented state form
Qm = np.zeros((param.nbVarX*param.nbData, param.nbVarX*param.nbData))

for t in range(param.nbData):
    id0 = np.linspace(0, param.nbVar-1, param.nbVar, dtype=int) + t * param.nbVar
    id = np.linspace(0, param.nbVarX-1, param.nbVarX, dtype=int) + t * param.nbVarX
    Qm[id[0]:id[-1]+1, id[0]:id[-1]+1] = np.vstack((np.hstack((np.identity(param.nbVar), np.zeros((param.nbVar, 1)))), np.append(-Mu[:, t].reshape(1, -1), 1))) \
    @ block_diag((Q[id0[0]:id0[-1]+1, id0[0]:id0[-1]+1]), 1) @ np.vstack((np.hstack((np.identity(param.nbVar), -Mu[:, t].reshape(-1, 1))), np.append(np.zeros((1, param.nbVar)), 1)))

Rm = np.identity((param.nbData-1)*param.nbVarU) * param.r


# Least squares formulation of recursive LQR with an augmented state space and and control primitives

xn = np.vstack((np.array([[3], [-0.5]]), np.zeros((param.nbVarX-param.nbVarU, 1))))  # Simulated noise on state

W = np.linalg.inv(PSI.T @ Su.T @ Qm @ Su @ PSI + PSI.T @ Rm @ PSI) @ PSI.T @ Su.T @ Qm @ Sx
F = PSI @ W  # F with control primitives

# Reproduction with feedback controller on augmented state space (with CP)
Ka = np.empty((param.nbData-1, param.nbVarU, param.nbVarX))
Ka[0, :, :] = F[0:param.nbVarU, :]
P = np.identity(param.nbVarX)
for t in range(param.nbData-2):
    id = t * param.nbVarU + np.linspace(2, param.nbVarU+1, param.nbVarU, dtype=int)
    P = P @ np.linalg.pinv(A - B @ Ka[t, :, :])
    Ka[t+1, :, :] = F[id, :] @ P

r = np.empty((2, param.nbVarX, param.nbData-1))
for n in range(2):
    x = np.append(Mu[:, 0] + np.append(np.array([2, 1]), np.zeros(param.nbVar-2)), 1).reshape(-1, 1)
    for t in range(param.nbData-1):
        # Feedback control on augmented state (resulting in feedback and feedforward terms on state)
        u = -Ka[t, :, :] @ x
        x = A @ x + B @ u  # Update of state vector
        if t == 24 and n == 1:
            x = x + xn  # Simulated noise on the state
        r[n, :, t] = x.flatten()  # State

# Plot 2D
plt.figure()
plt.axis("off")
plt.gca().set_aspect('equal', adjustable='box')

plt.plot(Mu[0, :], Mu[1, :], c='blue', linestyle='-', linewidth=2)
plt.scatter(Mu[0, -1], Mu[1, -1], c='red', s=100)
plt.scatter(r[0, 0, 0], r[0, 1, 0], c='black', s=50)
plt.plot(r[0, 0, :], r[0, 1, :], c='black', linestyle=':', linewidth=2)
plt.plot(r[1, 0, :], r[1, 1, :], c='black', linestyle='-', linewidth=2)
plt.scatter(r[1, 0, 23:25], r[1, 1, 23:25], c='green', s=30)

plt.show()